File: example/dnd.cpp
// Copyright (c) 2016, Herman Bergwerf. All rights reserved.
// Use of this source code is governed by a MIT-style license
// that can be found in the LICENSE file.

// Based on: http://seriss.com/people/erco/fltk/#DragAndDrop

#include <map>
#include <stdio.h>
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Box.H>

class Sender : public Fl_Box {
 public:
  Sender(int x, int y, int w, int h) : Fl_Box(x, y, w, h) {
    box(FL_FLAT_BOX);
    color(FL_GREEN);
    label("Drag from\nhere");
  }

  int handle(int event) {
    int ret = Fl_Box::handle(event);

    switch (event) {
    case FL_PUSH:
      Fl::copy("message", 7, 0);
      Fl::dnd();
      return 1;
    }

    return ret;
  }
};

class Receiver : public Fl_Box {
 public:
  Receiver(int x,int y,int w,int h) : Fl_Box(x,y,w,h) {
    box(FL_FLAT_BOX);
    color(FL_RED);
    label("to here");
  }

  int handle(int event) {
    int ret = Fl_Box::handle(event);

    switch (event) {
    case FL_DND_ENTER:
    case FL_DND_DRAG:
    case FL_DND_RELEASE:
      // return 1 for these events to 'accept' dnd
      return 1;

    case FL_PASTE:
      // handle actual drop (paste) operation
      label(Fl::event_text());
      return 1;
    }

    return ret;
  }
};

int main(int argc, char **argv) {
  // Create sender window and widget.
  auto win_a = new Fl_Window(0, 0, 200, 100, "Sender");
  auto a = new Sender(0, 0, 100, 100);
  win_a -> end();
  win_a -> show();

  // Create receiver window and widget.
  auto win_b = new Fl_Window(400, 0, 200, 100, "Receiver");
  auto b = new Receiver(100, 0, 100, 100);
  win_b -> end();
  win_b -> show();

  return Fl::run();
}

File: test/dnd.dart
// Copyright (c) 2016, Herman Bergwerf. All rights reserved.
// Use of this source code is governed by a MIT-style license
// that can be found in the LICENSE file.

import 'package:fltk/fltk.dart' as fl;

class Sender extends fl.Box {
  Sender(int x, int y, int w, int h) : super(x, y, w, h) {
    box = fl.FLAT_BOX;
    color = fl.GREEN;
    label = 'Drag from\nhere';
  }

  bool handle(fl.Event event) {
    switch (event) {
      case fl.Event.PUSH:
        fl.copy('message');
        fl.dnd();
        return true;

      default:
        return false;
    }
  }
}

class Receiver extends fl.Box {
  Receiver(int x, int y, int w, int h) : super(x, y, w, h) {
    box = fl.FLAT_BOX;
    color = fl.RED;
    label = 'to here';
  }

  bool handle(fl.Event event) {
    switch (event) {
      case fl.Event.DND_ENTER:
      case fl.Event.DND_DRAG:
      case fl.Event.DND_RELEASE:
        // Return true for these events to 'accept' dnd.
        return true;

      case fl.Event.PASTE:
        // Handle actual drop (paste) operation.
        label = fl.eventText;
        return true;

      default:
        return false;
    }
  }
}

int main() {
  // Create sender window and widget.
  final winA = new fl.Window.at(0, 0, 200, 100, 'Sender');
  new Sender(0, 0, 100, 100);
  winA.end();
  winA.show();

  // Create receiver window and widget.
  final winB = new fl.Window.at(400, 0, 200, 100, 'Receiver');
  new Receiver(100, 0, 100, 100);
  winB.end();
  winB.show();

  return fl.run();
}
